src/relentless/model/potential/bond.py

```diff
@@ -0,0 +1,568 @@
+        unique integer ID of the potential.
+
+    Attributes
+    ----------
+    coeff : :class:`BondParameters`
+        Parameters of the potential for each type.
+
+    Examples
+    --------
+    Kremer-Grest bond (no repulsion included)::
+
+        >>> u = relentless.potential.bond.FENEWCA(("A"))
+        >>> u.coeff["A"].update({'k': 30, 'r0': 1.5})
+
+    Kremer-Grest bond (repulsion included)::
+
+        >>> u = relentless.potential.bond.FENEWCA(("A"))
+        >>> u.coeff["A"].update({'k': 30, 'r0': 1.5, "epsilon":1, "sigma": 1})
+    """
+
+    def __init__(self, types, name=None):
+        super().__init__(keys=types, params=("k", "r0", "epsilon", "sigma"), name=name)
+
+    def energy(self, types, r):
+        """Evaluate bond energy."""
+        params = self.coeff.evaluate(types)
+        k = params["k"]
+        r0 = params["r0"]
+        epsilon = params["epsilon"]
+        sigma = params["sigma"]
+
+        # initialize arrays
+        r, FENE, s = self._zeros(r)
+        WCA = FENE.copy()
+
+        # set flags for FENE potential
+        fene_flag = ~numpy.greater_equal(r, r0)
+
+        # evaluate FENE potential
+        FENE[fene_flag] = -0.5 * k * r0**2 * numpy.log(1 - (r[fene_flag] / r0) ** 2)
+        FENE[~fene_flag] = numpy.inf
+
+        # set flags for WCA potential
+        zero_flags = ~numpy.isclose(r, 0)
+        wca_flags = r < 2 ** (1 / 6) * sigma
+
+        # evaluate WCA potential
+        r6_inv = numpy.power(sigma / r[zero_flags], 6)
+        WCA[zero_flags] = 4.0 * epsilon * (r6_inv**2 - r6_inv) + epsilon
+        WCA[~zero_flags] = numpy.inf
+        WCA[~wca_flags] = 0
+
+        if s:
+            FENE = FENE.item()
+            WCA = WCA.item()
+
+        return FENE + WCA
+
+    def force(self, types, r):
+        """Evaluate bond force."""
+        params = self.coeff.evaluate(types)
+        k = params["k"]
+        r0 = params["r0"]
+        epsilon = params["epsilon"]
+        sigma = params["sigma"]
+
+        # initialize arrays
+        r, dFENE_dr, s = self._zeros(r)
+        dWCA_dr = dFENE_dr.copy()
+
+        # set flags for FENE potential
+        fene_flag = ~numpy.greater_equal(r, r0)
+
+        # evaluate FENE potential
+        dFENE_dr[fene_flag] = (k * r0) / (1 - (r[fene_flag] / r0) ** 2)
+        dFENE_dr[~fene_flag] = numpy.inf
+
+        # set flags for WCA potential
+        zero_flags = ~numpy.isclose(r, 0)
+        wca_flags = r < 2 ** (1 / 6) * sigma
+
+        # evaluate WCA potential
+        rinv = 1.0 / r[zero_flags]
+        r6_inv = numpy.power(sigma * rinv, 6)
+        dWCA_dr[zero_flags] = (48.0 * epsilon * rinv) * (r6_inv**2 - 0.5 * r6_inv)
+        dWCA_dr[~zero_flags] = numpy.inf
+        dWCA_dr[~wca_flags] = 0
+
+        if s:
+            dFENE_dr = dFENE_dr.item()
+            dWCA_dr = dWCA_dr.item()
+
+        return dFENE_dr + dWCA_dr
+
+    def _derivative(self, param, r, k, r0, epsilon, sigma, **params):
```

Comment:
Same comments on this method regarding the flags. I did not check all the formulas below carefully, but I assume you have done them right in your unit tests / checked with Mathematica.